rfcs/0000-cosign-integration.md

```diff
@@ -0,0 +1,162 @@
+- It is unable to verify container image signature by digests.
+- Signature relocation is not possible with this tool.
+- It is a project that currently has low activity and community involvement,
+  since the community is currently focused on the development of Notary v2.
+
+[`cosign`](https://github.com/sigstore/cosign) presents itself as a solution to
+some of those aforementioned challenges. With `cosign`, the user is able to:
+- Relocate container images across registries along with its signatures, using
+  either:
+  - `crane cp` to copy images and signatures separately.
+  - `cosign copy` to copy images along with their signatures from one registry
+    to another.
+- Sign container images by tag or by digest.
+- Verify container image signatures by specifying either a tag or a digest.
+- Verify container image signatures after relocation.
+
+Moreover, `cosign` is a project that has had a lot of activity and community
+involvement recently.
+
+## Outcome
+
+`kpack` integrates with `cosign` to sign images it builds and push the
+signatures to a registry so that users can ensure the chain of custody of a
+generated artifact.
+
+This proposal aims to cover specifically the flow of signing an image produced
+by `kpack Image` resource without verification of base or builder images pulled
+in the process.
+
+## Actions to take
+
+### Enable image signing using `cosign`
+
+- Implement a flow that generates the signature payload for the image, then
+  calculates its signature and pushes it either to the registry where the image
+  is located, using the same credentials that were used to push the image, or
+  [to the registry specified in the `COSIGN_REPOSITORY` environment variable](#key-generation-and-storage).
+  This flow must happen after the image has been pushed to the registry.
+
+- If `cosign` fails to sign an image, the build should fail and output an error
+  message in the build log, so the operator can troubleshoot the issue. The
+  error messages should also be added in any other places where error messages
+  are presented.
+
+- Whenever `kpack` signs an image using `cosign`, it should add these
+  annotations:
+  - Build number.
+  - Build time.
+
+- Signing with `cosign` should not affect any configurations that enable signing
+  using the Notary v1 mechanism. These two signing mechanisms should be able to
+  coexist.
+
+### Key generation and storage
+
+`cosign` can use file-based keys, key management systems or hardware keys to
+ingest private keys, as well as a keyless flow that requires integration with
+[Rekor](https://github.com/sigstore/rekor).
+
+`kpack` should not generate the keys used for signing and verification. The user
+should pass them in using one of the mechanisms supported by `cosign`.
+In this RFC, we suggest that the first mechanism implemented be the use of
+Kubernetes `Secret`s for sending the relevant data into the build pod, by means
+of attaching secrets to the service account used by `kpack`.
```

Comment:
Could we clarify exactly how these secrets are used? Any build that's service account has the cosign secrets will be signed? What about builders?

Reply:
Yes, the proposal would be that any build that has the service account with such a secret attached would get signed. Our original idea was to leave out builders and sign images only. I'll update the text to make these parts clearer.

Reply:
I am a bit concerned that would require creating secrets with `cosign` and then altering them to support use in kpack. Would you object to being permissive and signing with any "cosign" secret?

Reply:
I think we should also sign builders if we are using the same secret inside service account mechanism for consistency.

Reply:
Could you clarify this is in the RFC?

Reply:
Sure thing! 😄

Reply:
I added a sentence in this section calling out the fact that any build that has a service account with cosign secrets attached to it will have signatures generated.

Reply:
Although I agree strongly that kpack should sign the builder images that it produces I also feel strongly that signing and verifying builder images is a second feature.

Signing end-user application images is the base feature that we have been discussing in this RFC so far.  I think that's the first feature we should implement.  kpack does not run these images so it does not need to be responsible for verifying signatures on them.  This RFC already covers this feature.

Signing builder images is a feature that I think we should approach after the base feature (signing application images) is complete.  This RFC does not yet discuss verifying signatures on any images.  It feels incomplete to me if one part of kpack is signing builder images that another part of kpack consumes, but does not verify the signatures on the builder images.

Can we please defer the signing and verifying of builder images to a second RFC?

Reply:
Implementing builder signing can definitely be implemented independently of individual image signing (and perhaps even released in different release). 

However, I believe it would still be valuable to specify how builder signing would work in this rfc. A benefit of the rfc process is that it enables us to design and evolve the api holistically at a higher level than individual features.  In this case particularly, allowing builders to be signed with the same mechanism as images is a big benefit of the proposed approach. It seems like a missed opportunity if we didn't provide a specify a consistent image signing approach for all the images that kpack produces.

Reply:
Builds scheduled with a kpack built builder image will be scheduled with the exact digest that kpack itself is reporting so, verifying them within kpack might not be necessary. However, A builder image may be used outside of kpack and that is where image signing seems valuable to me. If we want to verify builder images within kpack it may influence this design so, it might be useful to discuss.

Reply:
I updated the text to include signing of the builder image!